compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/AnnotationReference.kt

```diff
@@ -41,12 +51,46 @@ public sealed class AnnotationReference {
   public class Psi internal constructor(
     public val annotation: KtAnnotationEntry,
     override val classReference: ClassReference
-  ) : AnnotationReference()
+  ) : AnnotationReference() {
+
+    override fun isQualifier(): Boolean {
+      return annotation
+        .typeReference
+        ?.requireFqName(module)
+        // Often entries are annotated with @Inject, in this case we know it's not a qualifier and
+        // we can stop early.
+        ?.takeIf { it != injectFqName }
+        ?.classDescriptor(module)
+        ?.annotations
+        ?.hasAnnotation(qualifierFqName)
+        ?: false
```

Comment:
Same as for descriptors, you should be able to simplify that:
```suggestion
      return classReference.isAnnotatedWith(qualifierFqName)
```
This should work. If it does, then you don't even need an abstract function. The implementation for descriptors and PSI is the same thanks to `ClassReference` 😄